lib/active_merchant/billing/gateways/payu_latam.rb

```diff
@@ -0,0 +1,484 @@
+        else
+          message = "Unsupported operation: need valid order id to run a query."
+          return Response.new(false, message)
+        end
+
+        post[:details] = details
+
+        commit('query', post)
+      end
+
+      def supports_scrubbing?
+        true
+      end
+
+      def scrub(transcript)
+        transcript.
+          gsub(%r((\"creditCard\\\":{\\\"number\\\":\\\")\d+), '\1[FILTERED]').
+          gsub(%r((\"securityCode\\\":\\\")\d+), '\1[FILTERED]').
+          gsub(%r((\"apiKey\\\":\\\")\w+), '\1[FILTERED]')
+      end
+
+      private
+
+      def auth_or_sale(post, transaction_type, amount, payment_method, options)
+        add_credentials(post, 'SUBMIT_TRANSACTION')
+        add_transaction_type(post, transaction_type)
+        add_order(post, options)
+        add_buyer(post, options)
+        add_shipping_address(post, options)
+        add_invoice(post, amount, options)
+        add_signature(post)
+        add_payment_method(post, payment_method, options)
+        add_payer(post, options)
+        add_browser_data(post, options)
+        add_extra_parameters(post, options)
+      end
+
+      def add_credentials(post, command)
+        post[:test] = test? unless command == 'CREATE_TOKEN'
+        post[:language] = 'en'
+        post[:command] = command
+        merchant = {}
+        merchant[:apiLogin] = @options[:api_login]
+        merchant[:apiKey] = @options[:api_key]
+        post[:merchant] = merchant
+      end
+
+      def add_transaction_type(post, type)
+        transaction = {}
+        transaction[:type] = type
+        post[:transaction] = transaction
+      end
+
+      def add_order(post, options)
+        order = {}
+        order[:accountId] = @options[:account_id]
+        order[:referenceCode] = options[:order_id] || generate_unique_id
+        order[:description] = options[:description] || 'unspecified'
+        order[:language] = 'en'
+        post[:transaction][:order] = order
+      end
+
+      def add_buyer(post, options)
+        buyer = {}
+        buyer[:dniNumber] = options[:buyer_dni_number] if options[:buyer_dni_number]
+        buyer[:cnpj] = options[:buyer_cnpj] if options[:buyer_cnpj]
+        buyer[:emailAddress] = options[:email] || 'unspecified@example.com'
+        if address = options[:shipping_address]
+          buyer[:fullName] = address[:name]
+          shipping_address = {}
+          shipping_address[:street1] = address[:address1]
+          shipping_address[:street2] = address[:address2]
+          shipping_address[:city] = address[:city]
+          shipping_address[:state] = address[:state]
+          shipping_address[:country] = address[:country]
+          shipping_address[:postalCode] = address[:zip]
+          shipping_address[:phone] = address[:phone]
+          buyer[:shippingAddress] = shipping_address
+        end
+        post[:transaction][:order][:buyer] = buyer
+      end
+
+      def add_shipping_address(post, options)
```

Comment:
I know they sent an email about specifying the shipping address more than once.  I do wonder why that's needed!  Is there one that's more important than the other?  Seems weird to be doing it twice, once for the buyer and once for the transaction order.  Would it be bad to only specify it for the transaction order?